### PR TITLE
Support per-listener TLS configuration ⚙️

### PR DESCRIPTION
**Pull request by boxofrad**
This PR introduces the capability to configure TLS differently for Consul's listeners/ports (i.e. HTTPS, gRPC, and the internal multiplexed RPC port) which is useful in scenarios where you may want the HTTPS or gRPC interfaces to present a certificate signed by a well-known/public CA, rather than the certificate used for internal communication which must have a SAN in the form `server.<dc>.consul`.

**Example:**

```hcl
tls {
  defaults {
    verify_incoming = true
    cert_file = "path/to/internal/cert.pem"
    key_file = "path/to/internal/key.pem"
    ca_file = "path/to/internal/ca.pem"
  }

  internal_rpc {
    tls_min_version = "tls13"
  }

  https {
    cert_file = "path/to/https/cert.pem"
    key_file = "path/to/https/key.pem"
  }

  grpc {
    verify_incoming = false
    cert_file = "path/to/grpc/cert.pem"
    key_file = "path/to/grpc/key.pem"
  }
}
```

This change affects lots of files so I've split it roughtly into two commits: one making the requisite internal changes to the `tlsutil` package, and another changing the external configruation format. I also spent some time refactoring the `tlsutil` tests to deepen my understanding of the effects of the different settings, so that accounts for quite a big chunk of the diff.

For consistency with the version/cipher constants used to configure API Gateway we should revisit/rebase #11647 after this is merged. We'll also need to make some changes to support the auto-reload feature introduced in #12329.

### CODE REVIEW

File: agent/config/builder.go [hunk 1]
```diff
@@ -2473,3 +2452,75 @@ func validateAbsoluteURLPath(p string) error {
...
+	// so we call it out here too.
+	if t.Defaults.VerifyServerHostname != nil || t.GRPC.VerifyServerHostname != nil || t.HTTPS.VerifyServerHostname != nil {
+		return c, errors.New("verify_server_hostname is only valid in the tls.internal_rpc stanza")
+	}
+
+	// TLS is only enabled on the gRPC listener if there's an HTTPS port configured
+	// for historic and backwards-compatibility reasons.
+	if rt.HTTPSPort <= 0 && (t.GRPC != TLSListener{}) {
+		b.warn("tls.grpc was provided but TLS will NOT be enabled on the gRPC listener without an HTTPS listener configured (e.g. via ports.https)")
+	}
```
Comment (boxofrad) [inline agent/config/builder.go:2476]: For posterity: we considered decoupling the gRPC TLS "enabledness" from whether there's an HTTPS port configured, but it turns out there's a lot of subtle complexity hidden here (particularly when considering backwards compatibility and auto-config).

I thought we should at least warn the user if their `tls.grpc` settings will be ineffectual, because this tripped me up while testing!

**Comment by boxofrad** [comment]
Code changes are ready for review 🙇🏻 working on the docs changes now.

**Comment by boxofrad** [comment]
Docs changes done ✅

**Review by dhiaayachi — COMMENTED** [review]
Thank you for working on this @boxofrad!! I have a few questions but beside I'm now wondering if we should merge this and then I rework the auto-config reload PR accordingly?

File: tlsutil/config_test.go [hunk 2]
```diff
@@ -613,166 +1106,7 @@ func TestConfigurator_LoadCAs(t *testing.T) {
...
-	require.NoError(t, err)
-	tlsConf := c.commonTLSConfig(false)
-	require.Empty(t, tlsConf.CipherSuites)
-
-	conf := Config{CipherSuites: []uint16{
-		tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305}}
-	require.NoError(t, c.Update(conf))
-	tlsConf = c.commonTLSConfig(false)
-	require.Equal(t, conf.CipherSuites, tlsConf.CipherSuites)
-}
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config_test.go:657]: I think this is covered by the `CipherSuites` case in `TestConfigurator_IncomingConfig_Common`, which exercises it with a full handshake rather than checking the `tls.Config`.

https://github.com/hashicorp/consul/blob/dda057259133a625330c8f9eaafac8f8f78f5dca/tlsutil/config_test.go#L70-L91

You're talking about the CHACHA20_POLY1305 suites, right? I'm not sure if that value in the original test is relevant at all (or more likely randomly selected). I don't know enough about crypto stuff to say for sure, but I think you're right we should support them if Go does.

File: tlsutil/config_test.go [hunk 3]
```diff
@@ -613,166 +1106,7 @@ func TestConfigurator_LoadCAs(t *testing.T) {
...
-	require.Equal(t, c.commonTLSConfig(false).MinVersion, tlsLookup["tls10"])
-
-	for _, version := range tlsVersions() {
-		require.NoError(t, c.Update(Config{TLSMinVersion: version}))
-		require.Equal(t, c.commonTLSConfig(false).MinVersion,
-			tlsLookup[version])
-	}
-
-	require.Error(t, c.Update(Config{TLSMinVersion: "tlsBOGUS"}))
-}
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config_test.go:731]: Yep, this is covered by `TestConfigurator_IncomingConfig_Common` and `TestConfigurator_Validation`:

https://github.com/hashicorp/consul/blob/dda057259133a625330c8f9eaafac8f8f78f5dca/tlsutil/config_test.go#L47-L68

https://github.com/hashicorp/consul/blob/dda057259133a625330c8f9eaafac8f8f78f5dca/tlsutil/config_test.go#L869-L872

The default value comes from the `config` package's `DefaultSource`:

https://github.com/hashicorp/consul/blob/dda057259133a625330c8f9eaafac8f8f78f5dca/agent/config/default.go#L61

File: agent/config/runtime.go [hunk 4]
```diff
@@ -1711,23 +1711,40 @@ func (c *RuntimeConfig) Sanitized() map[string]interface{} {
...
-		Domain:                   c.DNSDomain,
-		ServerName:               c.ServerName,
-		TLSMinVersion:            c.TLSMinVersion,
-		CipherSuites:             c.TLSCipherSuites,
-		PreferServerCipherSuites: c.TLSPreferServerCipherSuites,
-		EnableAgentTLSForChecks:  c.EnableAgentTLSForChecks,
-		AutoTLS:                  c.AutoEncryptTLS || c.AutoConfig.Enabled,
+		InternalRPC: tlsutil.ListenerConfig{
+			VerifyIncoming:       c.VerifyIncoming || c.VerifyIncomingRPC,
+			VerifyOutgoing:       c.VerifyOutgoing,
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply agent/config/runtime.go]: This is just in the internal-only step change in [8b7d9d2](https://github.com/hashicorp/consul/pull/12504/commits/8b7d9d2a356c7fa3497d58de233703c4d8a25c44) and is removed with the changes to the external file format in [6db4696](https://github.com/hashicorp/consul/pull/12504/commits/6db469699dbb25cd400acc6415ad0da4352ec2ab) 🙇🏻‍♂️

File: tlsutil/config.go [hunk 5]
```diff
@@ -78,24 +110,21 @@ type Config struct {
...
 
-	// CertFile is used to provide a TLS certificate that is used for
-	// serving TLS connections.  Must be provided to serve TLS connections.
-	CertFile string
-
-	// KeyFile is used to provide a TLS key that is used for serving TLS
-	// connections.  Must be provided to serve TLS connections.
-	KeyFile string
+	// HTTPS is used to configure the external HTTPS listener.
+	HTTPS ListenerConfig
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config.go:127]: I don't feel strongly either way! The only gripe I have with the `Listener` naming is that in the case of Internal RPC (and to a smaller extent HTTPS) we also make _outgoing_ connections using this configuration, and "listener" doesn't feel right here, but I can't think of a better name (interface? socket? channel?) 😅

File: tlsutil/config.go [hunk 6]
```diff
@@ -368,46 +490,6 @@ func newX509CertPool(groups ...[]string) (*x509.CertPool, error) {
...
-		}
-
-		// We will use the auto_encrypt/auto_config cert for TLS in the incoming APIs when available. Therefore the check
-		// here will ensure that either we enabled one of those two features or a certificate and key were provided manually
-		if cert == nil && !config.AutoTLS {
-			return fmt.Errorf("VerifyIncoming requires either a Cert and Key pair in the configuration file, or auto_encrypt/auto_config be enabled")
-		}
-	}
-	return nil
-}
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config.go:405]: We certainly do. This validation has been moved into `loadListenerConfig`:

https://github.com/hashicorp/consul/blob/dda057259133a625330c8f9eaafac8f8f78f5dca/tlsutil/config.go#L290-L349

File: tlsutil/config_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikemorris) [reply tlsutil/config_test.go:657]: > I'm not sure if that value in the original test is relevant at all

It was somewhat relevant in that specifying that specific suite would (I believe) not pass through Consul's config [`ParseCiphers`](https://github.com/hashicorp/consul/blob/ee0e8bf50034de2b2ce13b933a1d6bf14c54403c/tlsutil/config.go#L947-L980) allow-list to be used in the actual TLS connection established - the test appeared to be testing a configuration that wouldn't actually work.

Would it be possible to add a test here for the negative case of specifying some older valid but insecure/unsupported cipher suite and ensure it _isn't_ used for the actual TLS connection? (I'm not sure if the correct behavior is for the connection to fail to be established, for it to error early, or for `CipherSuites` to fall back to defaults as if it was unspecified.)

[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config_test.go:657]: Oooooh I see, thanks for calling that out! Will add a test case to capture that 🙇🏻‍♂️

[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config_test.go:657]: Actually, having given this a bit more thought, I think we'd be testing that `CipherSuites` rejects the unsupported value. Which may be better served by a unit test on `CipherSuites` rather than a full-stack test like these — sorry I was a little slow on the uptake there! 😅

**Review by mkeeler — COMMENTED** [review]
My review of the code changes are complete. I still need to follow on with a review of the docs but I figured I would let loose the few comments I have so far.

File: agent/config/builder.go [hunk 7]
```diff
@@ -2473,3 +2452,75 @@ func validateAbsoluteURLPath(p string) error {
 
 	return nil
 }
+
+func (b *builder) buildTLSConfig(rt RuntimeConfig, t TLS) (tlsutil.Config, error) {
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply agent/config/builder.go:2456]: Yep, true! Although wouldn't this mean duplicating the relevant conversion logic from`builder.build`?

File: tlsutil/config.go [hunk 8]
```diff
@@ -44,31 +45,62 @@ var tlsLookup = map[string]uint16{
 	"tls13": tls.VersionTLS13,
 }
 
-// Config used to create tls.Config
-type Config struct {
+// ListenerConfig contains configuration for a given listener.
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config.go:48]: Definitely up for changing this! My only hesitation with "protocol" is that it makes me think of HTTP vs gRPC which is _sort of_ right except for the multiplexed port which supports multiple protocols. Maybe that's ok though, because it's more of a special case.

[the remark this answers is not in the document]
Comment (mkeeler) [reply tlsutil/config.go:48]: Technically our multiplex port is its own protocol with extra binary data in the stream. Its sort of like how HTTP is a protocol layered on TCP which is layered on IP etc. We have gRPC layered on the "Consul multiplexer protocol".

File: agent/config/builder.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (boxofrad) [reply agent/config/builder.go:2456]: Sure that's a possibility too, it's just that not all of that logic feels like it _belongs_ in this method (e.g. `DNSDomain` and `HTTPSPort`).

File: tlsutil/config.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (boxofrad) [reply tlsutil/config.go:48]: That's a great point! Ok, `ProtocolConfig` sounds good to me!

File: agent/config/builder.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/config/builder.go:2456]: good point, I think we can leave it as is then.

**Review by mkeeler — APPROVED** [review]

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/608579.
